design/deploy-steps.md

```diff
@@ -0,0 +1,158 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Customizable deployment procedure
+
+## Status
+
+implementable
+
+## Summary
+
+This design proposes customizing the deployment procedure by providing a high
+level abstraction around Ironic [deploy steps][deploy-steps], as well as
+allowing per-host deployment (IPA) images.
+
+[deploy-steps]: https://docs.openstack.org/ironic/latest/admin/node-deployment.html#deploy-steps
+
+## Motivation
+
+By default the Ironic agent downloads a provided image via HTTP(s), converts it
+(if needed) and writes directly to the disk. This procedure works for many
+distributions, but may not suit some cases well. A prominent example is CoreOS,
+which has its own installer (`coreos-installer`).
+
+### Goals
+
+- A way to replace the default deploy procedure with a custom action.
+- A way to provide a link to the Ironic agent image per host.
+
+### Non-Goals
+
+- Expose the full functionality of Ironic deploy steps.
+- Add a way to execute custom actions during deployment.
+- Building custom agent images.
+
+## Proposal
+
+### User Story
+
+As a CoreOS user I want to use `coreos-installer` to write the CoreOS image
+without losing any functionality that Ironic offers (unlike the live ISO
+workflow).
+
+## Design Details
+
+Additions to `BareMetalHost`:
+
+- New structure field `AgentImage` with the only field `URL`. It will accept
+  an optional link to a custom agent ISO image.
+- New structure field `CustomDeploy` with the only field `Method` to engage
+  a non-default deployment procedure as described below.
+- The `Image.URL` field will become optional if `CustomDeploy` is set
+  since we don't know if the deploy method actually requires a separate image.
+
+Setting `CustomDeploy` to a non-empty value will trigger deployment even if
+`Image.URL` is empty.
+
+### Implementation Details: AgentImage
+
+We will accept a custom agent ISO image to be used for each host in the new
+`AgentImage.URL` field (custom kernel/initramfs are out of scope for this
```

Comment:
If we're decoupling AgentImage and CustomDeploy APIs, do we expect AgentImage to also enable passing custom IPA images in the non-CustomDeploy case?

If not then I wonder if it should be one new spec field with two subfields, e.g something like:

```
  customDeploy:
    image: http://foo.iso
    method: custom-agent
```

Reply:
I think we do want to support custom IPA images for standard deployment to enable IPv4 static IPs and virtual media.

We could still put the agentImage field under the customDeploy section of the API.

Reply:
I like this option. IMO this makes a better user experience and explicit that the image being used as part of the custom deploy steps.

Reply:
> do we expect AgentImage to also enable passing custom IPA images in the non-CustomDeploy case

Why not? E.g. per-host network customization is not limited to CoreOS and may apply to a regular deployment the same way.

Reply:
> explicit that the image being used as part of the custom deploy steps

Exactly this impression I'm trying to avoid since it's not correct. These features are related, but are not mutually dependent.